### PR TITLE
Feat: Add padding to elements & minor fixes

### PR DESCRIPTION
**Pull request by AlexvZyl**
This PR adds a utility function to pad tables, as well as the option to pad the `packages` element.  If you like this I can add options to pad the other elements as well.

Also, there was a bug where the `shortcut` did not default properly.

No padding set (defaults to 1 above):
image

Padded 3 above and 2 bottom:
image

Reference: #272.

### CODE REVIEW

**Comment by glepnir** [comment]
LGTM need fix format

**Comment by AlexvZyl** [comment]
Will fix formatting, just want to add more padding features and optimize existing code.

**Comment by AlexvZyl** [comment]
Default setup:
image

Add 5 padding to each segment:
image

I now realize the centering does not take the padding into account.

File: lua/dashboard/theme/hyper.lua [hunk 1]
```diff
@@ -349,11 +355,6 @@ local function theme_instance(config)
     gen_footer(config)
     map_key(config, '<CR>')
     require('dashboard.events').register_lsp_root(config.path)
-    local size = math.floor(vim.o.lines / 2)
```
Comment (glepnir) [inline lua/dashboard/theme/hyper.lua:352]: why remove the auto pad at head

Comment (AlexvZyl) [reply lua/dashboard/theme/hyper.lua:352]: I am busy adding the option to manually pad the top.  And the existing code did not work well with the padding.

**Comment by AlexvZyl** [comment]
Would like me to add an auto centering function?

File: lua/dashboard/theme/hyper.lua (hunk 1 above)
Comment (AlexvZyl) [reply lua/dashboard/theme/hyper.lua:352]: > Default setup: image
> 
> Add 5 padding to each segment: image
> 
> I now realize the centering does not take the padding into account.

See here.

**Comment by glepnir** [comment]
yup . auto center better than add pad by manually.   pad in different section use pad option is correct.

File: lua/dashboard/theme/hyper.lua (hunk 1 above)
Comment (AlexvZyl) [reply lua/dashboard/theme/hyper.lua:352]: Will add centering that takes padding into account.

**Comment by AlexvZyl** [comment]
There is a big mistake in the current code.  Did not realize the `gen` functions are called inside a callback, so currently the padding gets added every time.  Will try to fix.

**Comment by glepnir** [comment]
the hyper  is run in luv call back .

**Comment by AlexvZyl** [comment]
Is there no initialization for the themes that gets called only once?  If not, would it be okay for me to add one?  Otherwise it might be difficult to setup the padding.

**Comment by glepnir** [comment]
it only init once . then the options will cache to local cache file. finally dashboard will clean all these data in memory make it clean.

**Comment by AlexvZyl** [comment]
Okay, I have an idea.  I am gonna make some changes, just yell at me if you do not like them.

**Comment by AlexvZyl** [comment]
Let me know what you think of that.  Padding now gets added only once.

File: lua/dashboard/init.lua [hunk 2]
```diff
@@ -182,13 +182,15 @@ function db:load_theme(opts)
     config = vim.tbl_extend('force', config, self.opts.preview)
   end
 
-  require('dashboard.theme.' .. opts.theme)(config)
+  local theme = require('dashboard.theme.' .. opts.theme)
+  theme.init(config)
```
Comment (glepnir) [inline lua/dashboard/init.lua:186]: does these will increase  startup time ?. header  use auto pad is enough like what i do. it add pad in callback after all content insert to buffer. then insert some `''` to the buffer head.

Comment (AlexvZyl) [reply lua/dashboard/init.lua:186]: Padding now only gets added once, whereas with the callbacks the header has to be regenerated with each callback.

As for the startup time, there is no difference (I think) for each method.  Both has to generate the header with padding at startup.

Comment (AlexvZyl) [reply lua/dashboard/init.lua:186]: I still only insert `''` as you requested, I am just making sure it only happens once.

Comment (glepnir) [reply lua/dashboard/init.lua:186]: okay . I will test it later.

**Comment by AlexvZyl** [comment]
(Still have to fix auto padding the header)

File: lua/dashboard/init.lua (hunk 2 above)
Comment (glepnir) [reply lua/dashboard/init.lua:186]: okay . I will test it later.

**Comment by glepnir** [comment]
just calculate the window height `vim.o.lines >> 1` and `api.nvim_buf_count() >> 1` when all lines insert.

**Comment by glepnir** [comment]
just calculate the window height `vim.o.lines >> 1` and `api.nvim_buf_line_count() >> 1` when all lines insert.

**Comment by AlexvZyl** [comment]
:)

image

File: lua/dashboard/theme/hyper.lua [hunk 3]
```diff
@@ -3,11 +3,11 @@ local utils = require('dashboard.utils')
...
 local function gen_shortcut(config)
-  local shortcut = vim.tbl_extend('force', {
+  local shortcut = config.shortcut or {
     { desc = '[  Github]', group = 'DashboardShortCut' },
     { desc = '[  glepnir]', group = 'DashboardShortCut' },
     { desc = '[  0.2.3]', group = 'DashboardShortCut' },
-  }, config.shortcut or {})
+  }
 
   if vim.tbl_isempty(shortcut) then
```
Comment (Prince213) [inline lua/dashboard/theme/hyper.lua:12]: This won't work properly as the code below assumes it has content. Consider returning early here.

**Review by Prince213 — CHANGES_REQUESTED** [review]

File: lua/dashboard/theme/hyper.lua [hunk 4]
```diff
@@ -3,11 +3,11 @@ local utils = require('dashboard.utils')
...
     shortcut = {}
```
Comment (Prince213) [inline lua/dashboard/theme/hyper.lua:13]:
```suggestion
    return
```

Comment (Prince213) [inline lua/dashboard/theme/hyper.lua:13]: This won't work properly. The code below still assumes it has content.

```suggestion
    return
```

**Review by Prince213 — CHANGES_REQUESTED** [review]

**Comment by AlexvZyl** [comment]
@glepnir If you still plan on merging this I can add padding for `Doom` as well.

**Comment by glepnir** [comment]
that's good. My thinking is that it makes sense to add pad options for different sections. But for the top I still want to use auto add pad. Provide as few options as possible to the user. Most people are not familiar with neovim api and they have people manually hardcode the pad values. Too many pad options can cause writing troubles for them.

**Comment by AlexvZyl** [comment]
Moving the cursor is currently broken with the `doom` theme.

**Comment by glepnir** [comment]
I saw you changed theme table to be like this. the design is what. export metatable looks wired. the pad options will in config table. then the config will pass to the `theme(config)`. can't simply use this way ?  btw thanks for your work

```lua
return {
  meta_table = meta_table,
  init = init
}
```

**Comment by AlexvZyl** [comment]
I did that so that the sections (header and footer) can be initialized once instead of being regenerated during runtime.  I did not change much, basically just added an `init(config)` function and then obviously the themes have to return a table so that both the function and meta table can be accessed.

**Comment by glepnir** [comment]
> can be initialized once

does this will take effect the `Dashbaord` command

**Comment by AlexvZyl** [comment]
This works when using `:Dashboard`.

`:Dashboard` -> `instance()` -> `load_theme()` -> `init()`.

However, I forgot that `:Dashboard` can be called by the user.  This means that when using the cached config, it will not pad correctly.  I will quickly fix that.

**Comment by AlexvZyl** [comment]
Fixed the padding when loading from cache.  Though, I do not know if you will like what I did.  Let me know.

**Comment by glepnir** [comment]
thanks i will test it later！

**Comment by AlexvZyl** [comment]
Hi @glepnir, do you still want this PR merged?  If you do, I will solve the merge conflicts.  If not, we can just close it.  Your call.

**Comment by cd-a** [comment]
why was this closed?

**Comment by AlexvZyl** [comment]
@glepnir Did a major refactor of dashboard after this PR was opened.  It would probably be easier to rewrite it.

**Pull request closed without merging**